lib/rom/reader.rb

```diff
@@ -58,6 +58,36 @@ def each
       mapper.process(relation) { |tuple| yield(tuple) }
     end
 
+    # Returns a single tuple from the relation if there is one.
+    #
+    # @raise [ROM::TupleCountMismatchError] if the relation contains more than
+    #   one tuple
+    #
+    # @api public
+    def one
+      if relation.count > 1
+        raise(
+          TupleCountMismatchError,
+          'The relation consists of more than one tuple'
+        )
+      else
+        mapper.process(relation).first
+      end
+    end
```

Comment:
For future reference: We discussed this on Gitter and we'll go with what we have right now, which means that `#one` and `#one!` currently require the relation to implement `#count`. In the future, we may provide a naive implementation using `Enumerator#next` (to check if there's more than one tuple in the relation) and the possibility to provide an implementation in the adapter in order to skip loading all records if it hurts performance.